File: app/src/main/java/com/candy/capture/core/SharedPreferenceManager.java
package com.candy.capture.core;

import android.content.Context;
import android.content.SharedPreferences;

/**
 * Created by zhanghq on 2016/11/11.
 */

public class SharedPreferenceManager {

    private static SharedPreferenceManager instance;
    private SharedPreferenceManager(Context context) {
        mContext = context;
        init();
    }

    public static SharedPreferenceManager getInstance(Context context) {
        if (instance == null) {
            synchronized (SharedPreferenceManager.class) {
                if (instance == null) {
                    instance = new SharedPreferenceManager(context.getApplicationContext());
                }
            }
        }
        return instance;
    }

    private Context mContext;
    private SharedPreferences mSharedPreferences;

    private String mLocationCity;
    private boolean mAllowFastCapture;
    private boolean mIsFirstRun;

    private static final String KEY_LOCATION_CITY = "key_location_city";
    private static final String KEY_FAST_CAPTURE = "key_fast_capture";
    private static final String KEY_IS_FIRST_RUN = "key_is_first_run";

    private void init() {
        mSharedPreferences = mContext.getSharedPreferences("capture", 0);

        mLocationCity = mSharedPreferences.getString(KEY_LOCATION_CITY, ConstantValues.DEFAULT_CITY_NAME);
        mAllowFastCapture = mSharedPreferences.getBoolean(KEY_FAST_CAPTURE, false);
        mIsFirstRun = mSharedPreferences.getBoolean(KEY_IS_FIRST_RUN, true);
    }

    public String getLocationCity() {
        return mLocationCity;
    }

    public void setLocationCity(String locationCity) {
        synchronized (SharedPreferenceManager.class) {
            if (!mLocationCity.equals(locationCity)) {
                mLocationCity = locationCity;
                mSharedPreferences.edit().putString(KEY_LOCATION_CITY, mLocationCity).apply();
            }
        }
    }

    public boolean isAllowFastCapture() {
        return mAllowFastCapture;
    }

    public void setAllowFastCapture(boolean allowFastCapture) {
        synchronized (SharedPreferenceManager.class) {
            mAllowFastCapture = allowFastCapture;
            mSharedPreferences.edit().putBoolean(KEY_FAST_CAPTURE, mAllowFastCapture).apply();
        }
    }

    public boolean isFirstRun() {
        return mIsFirstRun;
    }

    public void setFirstRun(boolean firstRun) {
        synchronized (SharedPreferenceManager.class) {
            mIsFirstRun = firstRun;
            mSharedPreferences.edit().putBoolean(KEY_IS_FIRST_RUN, mIsFirstRun).apply();
        }
    }
}

File: app/src/main/java/com/candy/capture/core/SharedPreferenceManager.kt
package com.candy.capture.core

import android.content.Context
import android.content.SharedPreferences
import androidx.datastore.DataStore
import androidx.datastore.preferences.Preferences
import androidx.datastore.preferences.createDataStore
import androidx.datastore.preferences.edit
import androidx.datastore.preferences.preferencesKey
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch

/**
 * @Description
 * @Author zhanghaiqiang
 * @Date 2020/9/30 11:24
 */
class SharedPreferenceManager private constructor(val context: Context) {

    companion object {
        private var instance: SharedPreferenceManager? = null

        fun getInstance(context: Context): SharedPreferenceManager {
            if (instance == null) {
                synchronized(SharedPreferenceManager::class.java) {
                    if (instance == null) {
                        instance = SharedPreferenceManager(context.applicationContext)
                    }
                }
            }
            return instance!!
        }
    }

    private var mLocationCity: String = ""
    private var mAllowFastCapture = false
    private var mIsFirstRun = false

    private val KEY_LOCATION_CITY = preferencesKey<String>("key_location_city")
    private val KEY_FAST_CAPTURE = preferencesKey<Boolean>("key_fast_capture")
    private val KEY_IS_FIRST_RUN = preferencesKey<Boolean>("key_is_first_run")

    private val dataStore: DataStore<Preferences> = context.createDataStore("capture")

    init {
        GlobalScope.launch {
            dataStore.data.collect {
                mLocationCity = it[KEY_LOCATION_CITY] ?: ConstantValues.DEFAULT_CITY_NAME
                mAllowFastCapture = it[KEY_FAST_CAPTURE] ?: false
                mIsFirstRun = it[KEY_IS_FIRST_RUN] ?: false
            }
        }
    }

    fun getLocationCity(): String? {
        return mLocationCity
    }

    fun setLocationCity(locationCity: String) {
        mLocationCity = locationCity
        GlobalScope.launch {
            dataStore.edit { settings ->
                settings[KEY_LOCATION_CITY] = locationCity
            }
        }
    }

    fun isAllowFastCapture(): Boolean {
        return mAllowFastCapture
    }

    fun setAllowFastCapture(allowFastCapture: Boolean) {
        mAllowFastCapture = allowFastCapture
        GlobalScope.launch {
            dataStore.edit { settings ->
                settings[KEY_FAST_CAPTURE] = allowFastCapture
            }
        }
    }

    fun isFirstRun(): Boolean {
        return mIsFirstRun
    }

    fun setFirstRun(firstRun: Boolean) {
        mIsFirstRun = firstRun
        GlobalScope.launch {
            dataStore.edit { settings ->
                settings[KEY_IS_FIRST_RUN] = firstRun
            }
        }
    }
}